radix-engine-common/src/data/scrypto/custom_well_known_types.rs

```diff
@@ -32,6 +34,87 @@ fn named_type_kind<L: SchemaTypeLink>(
     }
 }
 
+fn named_struct<L: SchemaTypeLink>(
+    name: &'static str,
+    fields: impl IntoIterator<Item = (&'static str, WellKnownTypeId)>,
+) -> ScryptoTypeData<L> {
+    TypeData::struct_with_named_fields(
+        name,
+        fields
+            .into_iter()
+            .map(|(name, field_type)| (name, field_type.into()))
+            .collect(),
+    )
+}
+
+fn named_tuple<L: SchemaTypeLink>(
+    name: &'static str,
+    fields: impl IntoIterator<Item = WellKnownTypeId>,
+) -> ScryptoTypeData<L> {
+    let field_types = fields
+        .into_iter()
+        .map(|field_type| field_type.into())
+        .collect();
+    TypeData::struct_with_unnamed_fields(name, field_types)
+}
+
+fn named_transparent<L: SchemaTypeLink>(
+    name: &'static str,
+    inner: ScryptoTypeData<L>,
+) -> ScryptoTypeData<L> {
+    inner.with_name(Some(name.into()))
+}
+
+fn array_of<L: SchemaTypeLink>(inner: WellKnownTypeId) -> ScryptoTypeData<L> {
+    TypeData {
+        kind: TypeKind::Array {
+            element_type: inner.into(),
+        },
+        metadata: TypeMetadata::unnamed(),
+        validation: TypeValidation::None,
+    }
+}
+
+fn named_enum<L: SchemaTypeLink>(
+    name: &'static str,
+    variants: impl IntoIterator<Item = (u8, ScryptoTypeData<L>)>,
+) -> ScryptoTypeData<L> {
+    TypeData::enum_variants(name, variants.into_iter().collect())
+}
+
+fn bytes_fixed_length_type_data<L: SchemaTypeLink>(length: usize) -> ScryptoTypeData<L> {
+    bytes_type_data().with_validation(TypeValidation::Array(LengthValidation {
+        min: Some(length.try_into().unwrap()),
+        max: Some(length.try_into().unwrap()),
+    }))
+}
+
+const REFERENCES_START: u8 = 0x80;
+const OWNED_ENTITIES_START: u8 = 0xa0;
+const MISC_TYPES_START: u8 = 0xc0;
+const KEY_TYPES_START: u8 = 0xd0;
+const ROLE_ASSIGNMENT_TYPES_START: u8 = 0xe0;
+const OTHER_MODULE_TYPES_START: u8 = 0xf0;
+
+//===============================================================================================================
+// SCRYPTO WELL KNOWN TYPES
+//===============================================================================================================
+// These correspond to either actual Scrypto value kinds, or other data types used in Scrypto / likely encountered
+// by builders writing blueprints, in their interfaces.
+//
+// There are a number of reasons for doing this:
+// 1. It is an optimization: Shared types mean smaller schemas.
+// 2. Having well known types for these so that these types are all the "same type" across all scrypto schemas.
```

Comment:
The only slight concern I have with this this "reason" is that we're making the execution/runtime dependent on high level notions of semantics. These are generally quite decoupled for separation of concerns and performance reasons.

Having said that, it's probably not a problem in this case since the overhead is quite low and can always be decoupled later on if it turns out to be a problem.